frontend/src/components/List/List.tsx

```diff
@@ -76,36 +81,91 @@ export class List extends React.Component<List.Props, {}> {
       // to rendering view, so we put the whole list in focus
       appState.nodes.setFocus(0, nodes.length);
     } else {
-      appState.nodes.setFocus(listStart, listEnd);
+      appState.nodes.setFocus(this.listStart, this.listEnd);
     }
 
     const height = TH_HEIGHT + nodes.length * TR_HEIGHT;
-    const transform = `translateY(${listStart * TR_HEIGHT}px)`;
+    const top = this.listStart * TR_HEIGHT;
+
+    nodes = nodes.slice(this.listStart, this.listEnd);
 
-    nodes = nodes.slice(listStart, listEnd);
+    const keys = this.recalculateKeys(nodes);
 
     return (
-      <React.Fragment>
+      <>
         <div className="List" style={{ height }}>
-          <table>
-            <Row.HEADER columns={selectedColumns} sortBy={sortBy} />
-            <tbody style={{ transform }}>
-              {nodes.map((node) => (
-                <Row
-                  key={node.id}
-                  node={node}
-                  pins={pins}
-                  columns={selectedColumns}
-                />
-              ))}
+          <table className="List--table">
+            <THead columns={selectedColumns} sortBy={sortBy} />
+            <tbody>
+              <tr className="List-padding" style={{ height: `${top}px` }} />
+              {nodes.map((node, i) => {
+                const key = keys[i];
+
+                this.previousKeys.set(node.id, key);
```

Comment:
I wonder if we could somehow avoid the churn on `this.previousKeys` and assign all the new keys at once. If the `unusedKeys` below retrieved the node id from `previousKey` too maybe we could avoid this here?

Reply:
Yeah, I think I can at least localize all the changes to `previousKeys` within `recalculateKeys`.